rooms.js

```diff
@@ -431,51 +432,144 @@ var GlobalRoom = (function () {
 				return;
 			}
 			newSearch.rating = mmr;
-			self.addSearch(newSearch, user);
+			self.addSearch(newSearch, user, formatid);
 		});
 	};
-	GlobalRoom.prototype.matchmakingOK = function (search1, search2, user1, user2) {
+	GlobalRoom.prototype.matchmakingOK = function (search1, search2, user1, user2, formatid) {
 		// users must be different
-		if (user1 === user2) return false;
+		if (user1 === user2) return null;
 
 		// users must have different IPs
-		if (user1.latestIp === user2.latestIp) return false;
+		if (user1.latestIp === user2.latestIp) return null;
 
-		// users must not have been matched immediately previously
-		if (user1.lastMatch === user2.userid || user2.lastMatch === user1.userid) return false;
+		// users must not have been matched recently
+		if (user1.lastSearchFoes.indexOf(user2.userid) >= 0 || user2.lastSearchFoes.indexOf(user1.userid) >= 0) return null;
 
 		// search must be within range
-		var searchRange = 100, formatid = search1.formatid, elapsed = Math.abs(search1.time - search2.time);
+		var searchRange = 100, elapsed = Math.abs(search1.time - search2.time);
 		if (formatid === 'ou' || formatid === 'oucurrent' || formatid === 'randombattle') searchRange = 50;
 		searchRange += elapsed / 300; // +1 every .3 seconds
 		if (searchRange > 300) searchRange = 300;
 		if (Math.abs(search1.rating - search2.rating) > searchRange) return false;
 
-		user1.lastMatch = user2.userid;
-		user2.lastMatch = user1.userid;
 		return true;
 	};
-	GlobalRoom.prototype.addSearch = function (newSearch, user) {
-		if (!user.connected) return;
-		for (var i = 0; i < this.searchers.length; i++) {
-			var search = this.searchers[i];
-			var searchUser = Users.getExact(search.userid);
-			if (!searchUser || !searchUser.connected) {
-				this.searchers.splice(i, 1);
-				i--;
-				continue;
+	GlobalRoom.prototype.addSearch = function (newSearch, user, formatid) {
+		// Filter racing conditions
+		if (!user.connected || user !== Users.getExact(user.userid)) return;
+		if (user.searching[formatid]) return;
+
+		if (!this.searches[formatid]) this.searches[formatid] = [];
+		var formatSearches = this.searches[formatid];
+		var totalSearches = formatSearches.length;
+		var rating = newSearch.rating;
+
+		// Figure out the placement using binary search
```

Comment:
I mean an scenario like this:
`[user1, user2, user3 ...]`

Where `user1` has just battled against `user2`. `user3` shares IP with `user1`. `user4` shares it too, because you know: school, and so does `user5`. We would need to try to match `user1` with all those players, in that priority, so performing an actual sorting would be a requirement (ok, not quite as complex as O(n^2) -I believe that's O(n log n) with an I-wanna-be-the-very-best implementation?).

I assume your non-naive without-caching O(n) implementation iterates over the list, and keeps track of exactly the two highest-rated players? That can't quite be done for the general case described above. Granted: it's unlikely, but so is the general issue described, and we'd better find a way to <i>completely</i> fix it.

```
Alternatively, we can just periodically match the player who's been waiting the longest. That's even faster (although still O(n) ) and will probably more noticeable than many other changes.
```

I prefer the caching-based solution.